plugin-scalebar/src/main/java/com/mapbox/pluginscalebar/ScaleBar.java

```diff
@@ -0,0 +1,91 @@
+package com.mapbox.pluginscalebar;
+
+import android.support.annotation.ColorRes;
+import android.support.annotation.NonNull;
+import android.support.annotation.UiThread;
+import android.support.annotation.VisibleForTesting;
+import android.view.View;
+
+import com.mapbox.mapboxsdk.camera.CameraPosition;
+import com.mapbox.mapboxsdk.maps.MapView;
+import com.mapbox.mapboxsdk.maps.MapboxMap;
+import com.mapbox.mapboxsdk.maps.Projection;
+
+/**
+ * Plugin class that shows a scale bar on MapView and changes the scale corresponding to the MapView's scale.
+ */
+public class ScaleBar {
+  private MapboxMap mapboxMap;
+  private Projection projection;
+  private boolean enabled = true;
+  private ScaleBarWidget scaleBarWidget;
+  private CameraMoveListener cameraMoveListener;
+
+  public ScaleBar(@NonNull MapView mapView, @NonNull MapboxMap mapboxMap) {
+    this.mapboxMap = mapboxMap;
+    this.projection = mapboxMap.getProjection();
+    this.scaleBarWidget = new ScaleBarWidget(mapView.getContext());
+    mapView.addView(scaleBarWidget);
+    cameraMoveListener = new CameraMoveListener();
+    mapboxMap.addOnCameraMoveListener(cameraMoveListener);
+    scaleBarWidget.setMapViewWidth(mapView.getWidth());
+  }
+
+  /**
+   * Returns true if the scale plugin is currently enabled and visible.
+   *
+   * @return true if enabled, false otherwise
+   */
+  public boolean isEnabled() {
+    return enabled;
+  }
+
+  /**
+   * Toggles the scale plugin state.
+   * <p>
+   * If the scale plugin wasn't enabled, a {@link com.mapbox.mapboxsdk.maps.MapboxMap.OnScaleListener}
+   * will be added to the {@link MapView} to listen to scale change events to update the state of this plugin. If the
+   * plugin was enabled the {@link com.mapbox.mapboxsdk.maps.MapboxMap.OnScaleListener} will be removed from the map.
+   * </p>
+   */
+  @UiThread
+  public void setEnabled(boolean enabled) {
+    if (this.enabled == enabled) {
+      // already in correct state
+      return;
+    }
+    this.enabled = enabled;
+    scaleBarWidget.setVisibility(enabled ? View.VISIBLE : View.GONE);
+    if (enabled) {
+      mapboxMap.addOnCameraMoveListener(cameraMoveListener);
+    } else {
+      mapboxMap.removeOnCameraMoveListener(cameraMoveListener);
+    }
+  }
+
+
+  /**
+   * Set colors for the scale bar.
+   *
+   * @param textColor      The color for the texts above the scale bar.
+   * @param primaryColor   The color for odd number index bars.
+   * @param secondaryColor The color for even number index bars.
+   */
+  public void setColors(@ColorRes int textColor, @ColorRes int primaryColor, @ColorRes int secondaryColor) {
+    scaleBarWidget.setColors(textColor, primaryColor, secondaryColor);
+  }
+
+  class CameraMoveListener implements MapboxMap.OnCameraMoveListener {
+    @Override
+    public void onCameraMove() {
+      CameraPosition cameraPosition = mapboxMap.getCameraPosition();
```

Comment:
The camera position should be cached, and `ScaleBarWidget#setDistancePerPixel` should only be called if `previousCamera.zoom != currentCamera.zoom` or `previousCamera.latitude != currentCamera.latitude`. Additionally, I think it'd be great to include thresholds, especially for latitude as it's not as sensitive as zoom level when it comes to the scale. For example, update only if `abs(previousCamera.latitude - currentCamera.latitude) > x`. This will help to limit the scale view updates even more.

Reply:
How can we choose this thresholds?

Reply:
@Chaoba you can verify how meters per pixel change based on latitude in our projection and select a threshold based on that. You will definitely need to update more frequently the closer you get to the poles.

Reply:
@LukasPaczos I have do a test and get the following conclusions:
1. `MetersPerPixelAtLatitude` changes more and more often when getting to the poles, and the logic will be complicated to choose these thresholds.
2. Currently  every time we zoom in/out or move camera, it will draw 1~2 times.

So IMHO, we shouldn't pay too much efforts to import so complicated logic for this tiny issue and take the risk of having potential bugs.

Reply:
Alright, thanks @Chaoba.
> Currently every time we zoom in/out or move camera, it will draw 1~2 times

I think this is acceptable, we can revisit with next iterations if needed.